src/metabase/pulse/render/body.clj

```diff
@@ -265,35 +267,75 @@
    "dddd, MMMM D, YYYY" {:week "MMMM D, YYYY"
                          :month "MMMM, YYYY"}})
 
+(defn- update-date-style
+  [date-style unit]
+  (let [unit (or unit :default)]
+    (or (get-in override-date-styles [date-style unit])
+        (get-in default-date-styles [unit])
+        date-style)))
+
+(defn- backfill-currency
+  [{:keys [number_style currency] :as settings}]
+  (cond-> settings
+    (and (= number_style "currency") (nil? currency))
+    (assoc :currency "USD")))
+
+(defn- settings-from-column
+  [col column-settings]
+  (or (get column-settings {::mb.viz/field-id (:id col)})
+      (get column-settings {::mb.viz/column-name (:name col)})))
+
+(defn- update-col-for-js
+  [col-settings col]
+  (-> (m/map-keys (fn [k] (-> k name (str/replace #"-" "_") keyword)) col-settings)
+      (backfill-currency)
+      (u/update-when :date_style update-date-style (:unit col))))
+
 (defn- ->js-viz
   "Include viz settings for js.
 
   - there are some date overrides done from lib/formatting.js
   - chop off and underscore the nasty keys in our map
   - backfill currency to the default of USD if not present"
   [x-col y-col {::mb.viz/keys [column-settings] :as _viz-settings}]
-  (letfn [(settings [col] (or (get column-settings {::mb.viz/field-id (:id col)})
-                              (get column-settings {::mb.viz/column-name (:name col)})))
-          (update-date-style [date-style unit]
-            (let [unit (or unit :default)]
-              (or (get-in override-date-styles [date-style unit])
-                  (get-in default-date-styles [unit])
-                  date-style)))
-          (backfill-currency [{:keys [number_style currency] :as settings}]
-            (cond-> settings
-              (and (= number_style "currency") (nil? currency))
-              (assoc :currency "USD")))
-          (for-js [col-settings col]
-            (-> (m/map-keys (fn [k] (-> k name (str/replace #"-" "_") keyword)) col-settings)
-                (backfill-currency)
-                (u/update-when :date_style update-date-style (:unit col))))]
-    (let [x-col-settings (settings x-col)
-          y-col-settings (settings y-col)]
-      (cond-> {:colors (public-settings/application-colors)}
-        x-col-settings
-        (assoc :x (for-js x-col-settings x-col))
-        y-col-settings
-        (assoc :y (for-js y-col-settings y-col))))))
+  (let [x-col-settings (settings-from-column x-col column-settings)
+        y-col-settings (settings-from-column y-col column-settings)]
+    (cond-> {:colors (public-settings/application-colors)}
+      x-col-settings
+      (assoc :x (update-col-for-js x-col-settings x-col))
+      y-col-settings
+      (assoc :y (update-col-for-js y-col-settings y-col)))))
+
+(defn- ->ts-viz
```

Comment:
I like more rigorous and typed settings. Hopefully we will have a migration and get rid of the js-viz function above. Is it possible to do this at once? I don't like having two different settings codepaths and hope it is short lived.

Reply:
the trouble is that the BE didn't actually have knowledge of the typed settings and now it does, but FE also only has partial knowledge of the typed settings and sometimes the typed settings are idiosyncratic to the individual static viz (cf. progress bars)

Reply:
> typed settings are idiosyncratic to the individual static viz (cf. progress bars)
then to me that means it is a bad name. If they are properly "progress-bar-settings" we should name them that rather than typed settings. But your response is just describing the current state right? Some parts use the new typed settings some don't. I'm hoping we can quickly migrate all over to one consistent view of the viz settings rather than having two different shapes (modulo particular bits needed for particular viz types). I'm just imagining in the future someone trying to reason if they need "js-viz" or "ts-viz" visualization settings.